### Importing libraries

In [1]:
import pandas as pd
import numpy as np
import os 

## Importing Data

In [2]:
# Set path
path = r'C:\Users\karya\Instacart Basket Analysis'

In [3]:
path

'C:\\Users\\karya\\Instacart Basket Analysis'

In [4]:
# Importing customer data
customer = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

### 4. Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.

In [5]:
customer.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [6]:
customer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [7]:
customer.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

## The Insights

* "First Name" needs to be changed to "first_name"
* "Surnam" needs to be changed to "surname"
* "Gender" needs to be changed to "gender"
* "STATE" needs to be changed to "state"
* "Age" needs to be change to "age"
*  I think "n_dependents" should be droped as it adds nothing to our analysis
* "user_id" can be used to join the dataframe with the "df_orders_products_merged" dataframe


### Renaming the Columns

#### To rename the column , run the code bellow:
df.rename(columns = {'original_col_name' : 'new_col_name'}, inplace = True)

In [8]:
# Renaming columns
customer.rename(columns = {'First Name':'first_name'}, inplace = True)
customer.rename(columns = {'Surnam':'surname'}, inplace = True)
customer.rename(columns = {'Gender':'gender'}, inplace = True)
customer.rename(columns = {'STATE':'state'}, inplace = True)
customer.rename(columns = {'Age':'age'}, inplace = True)

In [9]:
# Lets check updates
customer.head()

user_id first_name   surname  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

#### Q5. Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data

In [10]:
# Check for any mixed-type data
for col in customer.columns.tolist():
  weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customer[weird]) > 0:
    print (col)

C:\Users\karya\AppData\Local\Temp\ipykernel_7264\3589622709.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\karya\AppData\Local\Temp\ipykernel_7264\3589622709.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\karya\AppData\Local\Temp\ipykernel_7264\3589622709.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\karya\AppData\Local\Temp\ipykernel_7264\3589622709.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)


first_name


C:\Users\karya\AppData\Local\Temp\ipykernel_7264\3589622709.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\karya\AppData\Local\Temp\ipykernel_7264\3589622709.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\karya\AppData\Local\Temp\ipykernel_7264\3589622709.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\karya\AppData\Local\Temp\ipykernel_7264\3589622709.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\karya\AppData\L

#### The output above indicates that the column "first_name" contain some mixed-type data

In [11]:
# Checking for the NaN values
customer['first_name'].isna().sum()

11259

#### There are  11259  NaN values in the 'first_name' column. and in followin steps they will be imputed by "Unknown"


In [12]:
# Locating the NaN values 
customer.loc[customer['first_name'].isna()]

user_id first_name   surname  gender         state  age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        n_dependants fam_status  income  
53                 2    married   41709  
73                 0     single   82518  
82                 3    married  117099  
99                 1    married  155673  
105                3    married  151819  
...              ...        ...     ...  
206038             3    married   87783  
206044             2    married  108488  
206090             1    married   45275  
206162             0     single   56325  
206171             2    married   59222  

[11259 rows x 10 columns]

In [13]:
#Imputing the  NaN values with 'Unknown'
customer['first_name'].fillna('Unknown', inplace=True)

In [14]:
# Check again for any mixed-type data
for col in customer.columns.tolist():
  weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customer[weird]) > 0:
    print (col)

C:\Users\karya\AppData\Local\Temp\ipykernel_7264\2293347578.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\karya\AppData\Local\Temp\ipykernel_7264\2293347578.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\karya\AppData\Local\Temp\ipykernel_7264\2293347578.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\karya\AppData\Local\Temp\ipykernel_7264\2293347578.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (customer[[col]].applymap(type) != customer[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\karya\AppData\L

#### Checking the missing values:

In [15]:
customer.isnull().sum()

user_id         0
first_name      0
surname         0
gender          0
state           0
age             0
date_joined     0
n_dependants    0
fam_status      0
income          0
dtype: int64

#### The output [18] indicates that there is no missing value in the dataframe.

#### Checking for the duplicates

In [16]:
customer[customer.duplicated()]

Empty DataFrame
Columns: [user_id, first_name, surname, gender, state, age, date_joined, n_dependants, fam_status, income]
Index: []

### Q6. Combine your customer data with the rest of your prepared Instacart data. (Hint: Make sure the key columns are the same data type!)

In [17]:
# Importing  orders_products_merged 
orders_products_merged = pd.read_pickle(os.path.join( path, '02 Data', 'Prepared Data', 'orders_products_merged_updated_2.pkl'))

In [18]:
orders_products_customer_merged = orders_products_merged.merge(customer, on = ['user_id'])

In [19]:
orders_products_customer_merged.shape

(32404859, 34)

### Q8. Export this new dataframe as a pickle file so you can continue to use it in the second part of this task.

In [20]:
# Exporting the dataframe to pickle file
orders_products_customer_merged.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_customer_merged.pkl'))